### PR TITLE
RD-1769 Added support for new or modified filter

### PR DESCRIPTION
**Pull request by kubama**
image

### CODE REVIEW

File: widgets/filters/src/FilterModal.tsx [hunk 1]
```diff
@@ -24,13 +24,16 @@ const FilterModal: FunctionComponent<FilterModalProps> = ({
...
+
     const { useInput, useErrors } = Stage.Hooks;
     const [filterId, setFilterId] = useInput((initialFilter?.id ?? '') + initialFilterIdSuffix);
-    const [filterRules, setFilterRules] = useInput(
-        initialFilter ? [...initialFilter.attrs_filter_rules, ...initialFilter.labels_filter_rules] : []
-    );
+    const [filterRules, setFilterRules] = useInput(initialFilterRules);
     const { errors, setErrors, clearErrors, setMessageAsError } = useErrors();
-    const [filterRulesInvalid, setFilterRulesInvalid] = useState(false);
+    const [filterRulesInvalid, setFilterRulesInvalid] = useState(!initialFilter);
```
Comment (Gelio) [inline widgets/filters/src/FilterModal.tsx:36]: In `RulesForm` we have a `useEffect` that changes the rows when `initialFilters` change. I believe we should also change the `filterRulesInvalid` when the `initialFilter` changes.

Besides, would it not make more sense to have the `invalid` state be derived based on `filterRules` via `useMemo`? Then we would not need any manual synchronization

File: widgets/common/src/deploymentsView/useFilterQuery.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import { useQuery } from 'react-query';
+
+export default function useFilterQuery(toolbox: Stage.Types.Toolbox, filterId?: string) {
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/useFilterQuery.ts:3]: Good idea to extract it :+1:

File: widgets/common/src/deploymentsView/header/index.tsx [hunk 3]
```diff
@@ -28,15 +28,17 @@ const DeploymentsViewHeader: FunctionComponent<DeploymentsViewHeaderProps> = ({
     const [filterModalOpen, openFilterModal, closeFilterModal] = useBoolean();
     const [deployOnModalOpen, openDeployOnModal, closeDeployOnModal] = useBoolean();
     const [runWorkflowModalOpen, openRunWorkflowModal, closeRunWorkflowModal] = useBoolean();
-    const [filterId, setFilterId] = useState<string>();
+    const [userFilterRules, setUserFilterRules] = useState<FilterRule[]>();
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/index.tsx:31]: We have `userFilterRules` both in `DeploymentsView` and in here, as well as in `FilterModal`. Having duplicated state that needs to be synchronized could lead to problems and increased maintenance. Can we only have them in 2 places?
1. Rules that the widget uses to fetch the deployments
2. Local form state - the rules that the user sees

File: widgets/common/src/deploymentsView/DeploymentsView.tsx [hunk 4]
```diff
@@ -52,32 +52,11 @@ export const DeploymentsView: FunctionComponent<DeploymentsViewProps> = ({
...
-             * Such a case should not happen naturally, unless an `enabled` option is added to `useQuery`. If it is added,
-             * it should be here.
-             */
-            throw new Error('Idle state for fetching filter rules is not implemented.');
-        }
+    const [userFilterRules, setUserFilterRules] = useState<FilterRule[] | undefined>(undefined);
 
-        return result;
-    })();
+    const defaultFilterRulesResult = useFilterQuery(toolbox, userFilterRules ? undefined : defaultFilterId);
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/DeploymentsView.tsx:57]: `FilterModal` is already responsible for resolving the rules. Maybe we could also move the logic of resolving the initial filter rules there, so it will be the single point of resolving the filter rules? Then, it could propagate the state (`loading`, `error`, the actual rules) to the parent `DeploymentsView` component.

IMO that would make it easier to reason about, but I don't have strong arguments as I did not investigate the code deeply

File: widgets/common/src/deploymentsView/header/index.tsx [hunk 5]
```diff
@@ -50,20 +52,16 @@ const DeploymentsViewHeader: FunctionComponent<DeploymentsViewHeaderProps> = ({
...
+                        content={userFilterId ?? headerT('filter.unsavedFilter')}
                         onClick={openFilterModal}
                         style={{ whiteSpace: 'nowrap', maxWidth: 200, textOverflow: 'ellipsis', overflow: 'hidden' }}
                     />
-                    <Button
-                        icon="delete"
-                        onClick={() => handleFilterChange(undefined)}
-                        title={headerT('filter.clearButton')}
-                    />
+                    <Button icon="delete" onClick={() => handleFilterChange()} title={headerT('filter.clearButton')} />
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/index.tsx:64]: Minor:

I prefer seeing `undefined` passed explicitly rather than use default arguments  `handleFilterChange()`. I would make that function declaration look like:

```ts
function handleFilterChange(newFilterRules: FilterRule[] | undefined, newFilterId: string | undefined) {
```

which requires being explicit

File: widgets/common/src/deploymentsView/header/FilterModal.tsx [hunk 6]
```diff
@@ -56,15 +107,26 @@ const FilterModal: FunctionComponent<FilterModalProps> = ({
...
-                            onChange={handleChange}
+                            onChange={handleFilterIdChange}
                             fetchUrl="/filters/deployments?_include=id"
                             prefetch
                             value={filterId}
                         />
                     </UnsafelyTypedFormField>
+                    <UnsafelyTypedFormField label={tModal('filterRules')}>
+                        {filterRulesResult.isFetching && <Stage.Basic.LoadingOverlay />}
+                        {filterRulesResult.isSuccess && (
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/FilterModal.tsx:121]: What about `filterRulesResult.isError` :smile: ?

File: widgets/common/src/deploymentsView/header/FilterModal.tsx [hunk 7]
```diff
@@ -21,33 +26,79 @@ const FilterModal: FunctionComponent<FilterModalProps> = ({
...
-    const [filterId, setFilterId, resetFilterId] = useResettableState<string | undefined>(filterIdProp);
+    const [submitedFilterId, setSubmitedFilterId] = useState<string>();
+    const [filterId, setFilterId] = useState<string>();
+    const [initialFilterRules, setInitialFilterRules] = useState<FilterRule[]>([]);
+    const [filterRules, setFilterRules] = useState<FilterRule[]>();
+    const [rulesErrorsPresent, setRuleErrorsPresent] = useState<boolean>();
+
+    // Controlls filter 'dirty' state
+    const [filterEdited, setFilterEdited, unsetFilterEdited] = useBoolean();
+    const [submitedFilterEdited, setSubmitedFilterEdited] = useState<boolean>();
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/FilterModal.tsx:40]: Typo: submited -> submitted

**Review by Gelio — COMMENTED** [review]
The structure looks cool :+1: I did not test it in the UI yet.

Do you plan on running a full run of system tests for this PR?

I see you're targeting the `6.1.0-build` branch with this PR. Do you know if we already switched the CI to build off this branch? I am not aware of any such changes that have happened so far

File: widgets/common/src/deploymentsView/DeploymentsView.tsx (hunk 4 above)
Comment (kubama) [reply widgets/common/src/deploymentsView/DeploymentsView.tsx:57]: My understanding is that there is no direct correlation between the default (configured) filter and filter selection modal, so I would actually consider such solution artificial. The extracted `useFilterQuery` is a common hook, there is almost no extra logic needed to get the filter rules, so I consider the current solution sufficient.

File: widgets/common/src/deploymentsView/header/FilterModal.tsx (hunk 7 above)
Comment (kubama) [reply widgets/common/src/deploymentsView/header/FilterModal.tsx:40]: Done

File: widgets/common/src/deploymentsView/header/index.tsx (hunk 5 above)
Comment (kubama) [reply widgets/common/src/deploymentsView/header/index.tsx]: Done

File: widgets/common/src/deploymentsView/header/FilterModal.tsx (hunk 6 above)
Comment (kubama) [reply widgets/common/src/deploymentsView/header/FilterModal.tsx:126]: I added `useEffect` that displays an error message when `filterRulesResult.isError` flag gets set

File: widgets/common/src/deploymentsView/header/index.tsx (hunk 3 above)
Comment (kubama) [reply widgets/common/src/deploymentsView/header/index.tsx:31]: This state value is present in 3 places but it is used in each of them:
1. Top level - `header/FilterModal` - local form state, value is read and written
2. Middle level - `header/index` - used to set proper button label, but the value is also unset from here when the user filter is removed, so again the value is read and written
3. Bottom level - `DeploymentsView` - rules that the widget uses to fetch the deployments, read only

Summing up, I don't see an easy way to remove it from the second level without introducing one or more other state variables in place.

File: widgets/filters/src/FilterModal.tsx (hunk 1 above)
Comment (kubama) [reply widgets/filters/src/FilterModal.tsx:36]: > In `RulesForm` we have a `useEffect` that changes the rows when `initialFilters` change. I believe we should also change the `filterRulesInvalid` when the `initialFilter` changes.

It's a different component, so I'm not sure what's the reasoning. Besides, by design the `!initialFilter` value never changes - it's constant for each of three modal types (add/edit/clone).

> Besides, would it not make more sense to have the `invalid` state be derived based on `filterRules` via `useMemo`? Then we would not need any manual synchronization

The `filterRulesInvalid` is not just a derived value, it is an actual state variable that gets mutated.

**Comment by kubama** [comment]
> Do you plan on running a full run of system tests for this PR?

I added related system tests to this PR, scheduled full build: https://jenkins.cloudify.co/view/UI/job/Stage-UI-System-Test/816/

**Comment by Gelio** [comment]
Minor:

I have noticed that when there is a filter set, there is no gap between the blue filter button and "Bulk Actions":

image

**Comment by kubama** [comment]
Another system tests run after correcting flaky test: https://jenkins.cloudify.co/view/UI/job/Stage-UI-System-Test/821/

**Comment by kubama** [comment]
> I have noticed that when there is a filter set, there is no gap between the blue filter button and "Bulk Actions"

Thanks for noticing, fixed.

**Comment by kubama** [comment]
Looks like the build is failing due to https://github.com/cloudify-cosmo/docs.getcloudify.org/pull/1880 being merged and related README not updated.

File: widgets/common/src/deploymentsView/header/index.tsx (hunk 3 above)
Comment (Gelio) [reply widgets/common/src/deploymentsView/header/index.tsx:31]: I ended up drawing as I would do that myself and I ended up with a very similar structure to yours:

image

So, looks like I was wrong initially and this is the way :slightly_smiling_face:

File: widgets/common/src/deploymentsView/DeploymentsView.tsx (hunk 4 above)
Comment (Gelio) [reply widgets/common/src/deploymentsView/DeploymentsView.tsx:57]: Yup, after drawing it myself, looks ok :+1:

Comment (Gelio) [reply widgets/common/src/deploymentsView/DeploymentsView.tsx:57]: I can't open a new conversation for some reason, so I will use this one.

Minor:

Do we need this ternary operator for the 2nd argument? IMO this could always point to `defaultFilterId`. It's just the `defaultFilterRulesResult` won't be used later on when there are used defined filter rules.

Otherwise, it may happen that the user will see a loading/error state even though they just filled in the form and are using user-defined rules

**Review by Gelio — COMMENTED** [review]
As I am reading the code, I find the logic complex :worried: There's a lot of similar state, unfortunately, and that makes it more confusing. I don't have any idea how to simplify it, as I would have to probably try to implement something similar myself.

One idea may be to use `useReducer` to simplify state management for similar pieces of state. Maybe that would centralize the state management instead of using individual multiple `set...` calls one after another (in `header/index` and `header/FilterModal`)

I did not read the tests and I do not fully comprehend the `FilterModal`. I think I will only review the tests next (later today) and trust you that `FilterModal` is fine, because frankly, the code is complex to me and reading it takes too much of my time (I have spent ~2.5h for this PR already)

File: widgets/filters/src/FilterModal.tsx (hunk 1 above)
Comment (kubama) [reply widgets/filters/src/FilterModal.tsx:36]: > Right, I did not notice this is a separate component. I was confused because this was changed in the PR that I believe was scoped only for Deployments View.
> 
> Anyway, a minor risk I see now is that we have this `initialFilter`/`initialFilters` props that are sometimes reactive (a change will update the component's internal state) and sometimes they don't.
> 
> As this is not related to this PR (as is pretty much this whole change, if I'm not mistaken 😄 ), we can tackle it later, in some separate ticket. We can keep this discussion going to clarify it, but it does not block the PR 🙂

The changes here are related to the PR. Please note the change in `RulesForm` - I added an effect to update form content on `initialFilters` change. I needed this to implement form reloading when changing the filter selected in the dropdown. All changes in this `FilterModal` component are a result of this change and they are intended to preserve the pre-existing behaviour.

> Right, I can see that 🙂 However, from the logical perspective, the fact that the form is valid or not, to my mind, depends on the form's content, meaning the rules. Thus, if we make it a separate state variable, we run the risk of desynchronizing the state in some case in which we only change the rules and forget to update the validity. Looks like a risk to me and higher maintenance cost 🙂
> 
> Is there a case when we need to mutate the validity so that it is inconsistent with the form rules?

The `RulesForm` component reports form validity as part of the `onChange` callback. However it is called only as a result of user interaction with the form. When the form is in its initial state the `onChange` is not yet called and the form is valid if it's prepopulated with existing filter rules and invalid when it is empty. Hence the state variable with the initial value. The variable is always in sync with the form state and there is no way it could be inconsistent, as it is only updated as part of the `onChange` callback.

File: widgets/filters/src/FilterModal.tsx [hunk 8]
```diff
@@ -24,13 +24,16 @@ const FilterModal: FunctionComponent<FilterModalProps> = ({
     onSubmit,
     toolbox
 }) => {
+    const initialFilterRules = useMemo(
+        () => (initialFilter ? [...initialFilter.attrs_filter_rules, ...initialFilter.labels_filter_rules] : []),
+        []
```
[the remark this answers is not in the document]
Comment (kubama) [reply widgets/filters/src/FilterModal.tsx:29]: `initialFilterRules` will never become stale or outdated. By design this modal is always opened in the context of a single filter. `initialFilter` is always passed as a static data, there is no async loading involved. In other words there is no scenario in which `initialFilter` would ever change.

File: widgets/filters/src/FilterModal.tsx [hunk 9]
```diff
@@ -24,13 +24,16 @@ const FilterModal: FunctionComponent<FilterModalProps> = ({
     onSubmit,
     toolbox
 }) => {
+    const initialFilterRules = useMemo(
```
[the remark this answers is not in the document]
Comment (kubama) [reply widgets/filters/src/FilterModal.tsx:27]: Please see https://github.com/cloudify-cosmo/cloudify-stage/pull/1444#discussion_r656048177

File: widgets/common/src/deploymentsView/header/FilterModal.tsx [hunk 10]
```diff
@@ -21,33 +27,83 @@ const FilterModal: FunctionComponent<FilterModalProps> = ({
...
     const { errors, setErrors, clearErrors } = Stage.Hooks.useErrors();
-    const [filterId, setFilterId, resetFilterId] = useResettableState<string | undefined>(filterIdProp);
+    const [submittedFilterId, setSubmittedFilterId] = useState<string>();
+    const [filterId, setFilterId] = useState<string>();
+    const [initialFilterRules, setInitialFilterRules] = useState<FilterRule[]>([]);
+    const [filterRules, setFilterRules] = useState<FilterRule[]>();
+    const [rulesErrorsPresent, setRuleErrorsPresent] = useState<boolean>();
+
+    // Controlls filter 'dirty' state
+    const [filterEdited, setFilterEdited, unsetFilterEdited] = useBoolean();
```
[the remark this answers is not in the document]
Comment (kubama) [reply widgets/common/src/deploymentsView/header/FilterModal.tsx]: Right, renamed

File: widgets/filters/src/FilterModal.tsx (hunk 1 above)
Comment (Gelio) [reply widgets/filters/src/FilterModal.tsx:36]: > I needed this to implement form reloading when changing the filter selected in the dropdown. All changes in this `FilterModal` component are a result of this change and they are intended to preserve the pre-existing behaviour.

Makes sense now :+1: 

For the other part - maybe there could be some pattern to make it clearer and avoid so many interactions between components, but I did not come up with one just by reviewing. Let's leave it be

File: widgets/filters/src/FilterModal.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (Gelio) [reply widgets/filters/src/FilterModal.tsx:29]: > `initialFilter` is always passed as a static data, there is no async loading involved. In other words there is no scenario in which `initialFilter` would ever change.

We're making this assumption implicitly :slightly_smiling_face: Maybe you know it after reading the interaction between a few components, but just by looking at this code, it's unexpected - you have a variable that is used in the callback, but not used in the dependency array. This is a basic violation of rules of hooks and would be reported as such by the ESLint plugin for rules of hooks :slightly_smiling_face: I agree that it will likely not change, but there is no harm in adding the `initialFilter` to the dependency array. Then, we are more compliant with how `useMemo` is used.

If you are not too keen on modifying the dependency array, a comment stating what you just said (that `initialFilter` is static and will never change) makes this clear enough.

I would advise going with the way of modifying the dependency array anyway

File: widgets/common/src/filters/RulesForm.tsx [hunk 11]
```diff
@@ -56,23 +56,33 @@ export interface RulesFormProps {
...
         );
+        setRows(newRows);
+        handleChange(newRows);
     }
 
     useEffect(() => {
-        onChange(rows.map(getFilterRule), rows.filter(row => row.hasError).length > 0);
-    }, [rows]);
+        setRows(getFilterRuleRows(initialFilters));
+    }, [initialFilters]);
```
Comment (Gelio) [inline widgets/common/src/filters/RulesForm.tsx:85]: `RulesForm` is a controlled and an uncontrolled element at the same time, which is probably why I find the solution confusing. It has its own state, but also takes state from above, and communicated the changes to parent components. I am ok with having it here for now, but I would definitely try to find some better approach in the future - maybe centralized `useReducer` in some parent component or some advanced React patterns

File: widgets/common/src/deploymentsView/header/FilterModal.tsx [hunk 12]
```diff
@@ -21,33 +27,83 @@ const FilterModal: FunctionComponent<FilterModalProps> = ({
     const { ApproveButton, CancelButton, Icon, Modal, UnsafelyTypedForm, UnsafelyTypedFormField } = Stage.Basic;
     // @ts-expect-error DynamicDropdown is not converted to TS yet
     const { DynamicDropdown } = Stage.Common;
-    const { useResettableState } = Stage.Hooks;
+    const { useBoolean } = Stage.Hooks;
 
     const { errors, setErrors, clearErrors } = Stage.Hooks.useErrors();
-    const [filterId, setFilterId, resetFilterId] = useResettableState<string | undefined>(filterIdProp);
+    const [submittedFilterId, setSubmittedFilterId] = useState<string>();
+    const [filterId, setFilterId] = useState<string>();
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/FilterModal.tsx:34]: Did you consider my previous suggestion about using `useReducer`?

File: test/cypress/integration/widgets/deployments_view_spec.ts [hunk 13]
```diff
@@ -253,14 +241,24 @@ describe('Deployments View widget', () => {
...
             cy.contains(deploymentNameThatMatchesFilter);
             cy.contains(deploymentName);
-            cy.interceptSp('POST', '/searches/deployments').as('deploymentsSearchRequest');
 
-            widgetHeader.setFilter(filterId);
+            cy.contains('button', 'Filter').click();
+
+            cy.get('.modal').within(() => {
+                cy.setSearchableDropdownValue('Filter ID', filterId);
+                cy.contains('Label').should('be.visible');
```
Comment (Gelio) [inline test/cypress/integration/widgets/deployments_view_spec.ts:254]: This is unexpected to me. Which "Label" are we looking for? Could you add some comment?

**Review by Gelio — APPROVED** [review]
Looks ok to me, but it still looks complex :worried: I did not fully comprehend all the changes, but I don't want to spend more time digging into the details, as I have already spent quite some time on this PR

Since this is quite a substantial change, I believe @stepek should also take a look to have 2 approvals

File: test/cypress/integration/widgets/deployments_view_spec.ts (hunk 13 above)
Comment (kubama) [reply test/cypress/integration/widgets/deployments_view_spec.ts]: I extracted a function, I think it should be obvious now.

**Comment by kubama** [comment]
> As I am reading the code, I find the logic complex 😟 There's a lot of similar state, unfortunately, and that makes it more confusing. I don't have any idea how to simplify it, as I would have to probably try to implement something similar myself.
> 
> One idea may be to use `useReducer` to simplify state management for similar pieces of state. Maybe that would centralize the state management instead of using individual multiple `set...` calls one after another (in `header/index` and `header/FilterModal`)

One reason it looks complex is that except 'current' modal state we also have 'submitted' state, which is saved to be able to revert to it when closing the modal with cancel. I took advantage of this pattern and created `useRevertableState` hook which I reused for handling `filterId`, `filterRules` and `filterDirty` state values. I think it's much cleaner now and also easier to understand. The only confusing thing now is the `initialFilterRules` state variable which is needed only because, as you noted, `RulesForm` is not a fully controlled component at the moment. I'm not willing to convert it at this point as its internal state is derived from the initial state provided through the prop, so the conversion wouldn't be straightforward - looks like for each change there would need to be a conversion from internal state to external data representation and vice versa.

File: widgets/common/src/deploymentsView/header/FilterModal.tsx (hunk 12 above)
Comment (kubama) [reply widgets/common/src/deploymentsView/header/FilterModal.tsx]: Yes, I considered it but decided to take a different path - please see https://github.com/cloudify-cosmo/cloudify-stage/pull/1444#issuecomment-866082811

File: widgets/filters/src/FilterModal.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (kubama) [reply widgets/filters/src/FilterModal.tsx]: OK, updated

**Comment by kubama** [comment]
I'm not planning any more changes at this point. So I'm waiting for a final review.

@stepek Please let me know if you are willing to submit your review.

File: widgets/filters/src/FilterModal.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (Gelio) [reply widgets/filters/src/FilterModal.tsx]: Thanks! :+1:

File: test/cypress/integration/widgets/deployments_view_spec.ts (hunk 13 above)
Comment (Gelio) [reply test/cypress/integration/widgets/deployments_view_spec.ts]: Yup, much better :smile:

File: widgets/common/src/deploymentsView/header/index.tsx [hunk 14]
```diff
@@ -28,15 +28,15 @@ const DeploymentsViewHeader: FunctionComponent<DeploymentsViewHeaderProps> = ({
     const [filterModalOpen, openFilterModal, closeFilterModal] = useBoolean();
     const [deployOnModalOpen, openDeployOnModal, closeDeployOnModal] = useBoolean();
     const [runWorkflowModalOpen, openRunWorkflowModal, closeRunWorkflowModal] = useBoolean();
-    const [userFilterRules, setUserFilterRules] = useState<FilterRule[]>();
+    const [userFilterSelected, setUserFilterSelected] = useState<boolean>(false);
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/index.tsx:31]: Good simplification :+1:

File: widgets/common/src/deploymentsView/header/FilterModal.tsx [hunk 15]
```diff
@@ -16,8 +16,37 @@ interface FilterModalProps {
...
+    }
+
+    const result: [T, typeof setValue, typeof saveValue, typeof revertValue, typeof resetValues] = [
+        value,
+        setValue,
+        saveValue,
+        revertValue,
+        resetValues
+    ];
+    return result;
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/FilterModal.tsx:45]: Minor:

This looks like a bit too many values in "tuple" to me :slightly_smiling_face: If we returned an object with properties `value`, `set`, `save, `revert`, `reset`, we could use it like:

```ts
const submittedFilterIdState = useRevertableState(...);

submittedFilterIdState.value;
submittedFilterIdState.set('abc');
```

IMO less typing for us and fewer places to mistakes when using the hook (with an object it's harder to use the wrong function, as they are explicitly named when exported)

File: widgets/common/src/deploymentsView/header/FilterModal.tsx [hunk 16]
```diff
@@ -16,8 +16,37 @@ interface FilterModalProps {
 const tModal = Stage.Utils.getT(`${i18nPrefix}.header.filter.modal`);
 const tMessage = Stage.Utils.getT(i18nMessagesPrefix);
 
+function useRevertableState<T>(initialValue: T) {
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/FilterModal.tsx:19]: I like the idea :+1: Also, good work keeping this hook local to the component and not extracting it preemptively - I like it

File: widgets/common/src/deploymentsView/header/FilterModal.tsx [hunk 17]
```diff
@@ -27,27 +56,33 @@ const FilterModal: FunctionComponent<FilterModalProps> = ({
...
+    // The values are 'saved' on modal submit and 'reverted' on modal cancel
+    const [filterId, setFilterId, saveFilterId, revertFilterId, resetFilterId] = useRevertableState<string | undefined>(
+        undefined
+    );
+    const [filterRules, setFilterRules, saveFilterRules, revertFilterRules, resetFilterRules] = useRevertableState<
+        FilterRule[]
+    >([]);
+    const [filterDirty, setFilterDirty, saveFilterDirty, revertFilterDirty, resetFilterDirty] = useRevertableState<
+        boolean
+    >(false);
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/FilterModal.tsx:72]: Minor:

Maybe we could extract these 3 to a custom hook and expose functions to work with the form in an easier way? I see they are often called together, so maybe we could group this "call together" logic into functions inside of a custom hook

**Review by Gelio — DISMISSED** [review]
Looks good :+1: The `useRevertableState` hook cleared some stuff up, nice idea :smile: 

I have suggested some possible improvements, but they are not a must-have and could be postponed if you think they are worth it :slightly_smiling_face:

File: widgets/common/src/deploymentsView/header/FilterModal.tsx (hunk 15 above)
Comment (kubama) [reply widgets/common/src/deploymentsView/header/FilterModal.tsx]: Great idea, implemented.

File: widgets/common/src/deploymentsView/header/FilterModal.tsx (hunk 17 above)
Comment (kubama) [reply widgets/common/src/deploymentsView/header/FilterModal.tsx]: I implemented the idea, probably not in the most optimal way though - please review.

File: widgets/common/src/deploymentsView/header/FilterModal.tsx [hunk 18]
```diff
@@ -22,27 +22,24 @@ function useRevertableState<T>(initialValue: T) {
...
+        save: () => setSavedValue(value),
+        revert: () => setValue(savedValue),
+        reset: () => {
+            resetValue();
+            resetSavedValue();
+        }
+    };
+}
+
+function useRevertableStates(...revertableStates: Omit<ReturnType<typeof useRevertableState>, 'set'>[]) {
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/FilterModal.tsx:37]: Nice TS magic :smile:

File: widgets/common/src/deploymentsView/header/FilterModal.tsx [hunk 19]
```diff
@@ -61,31 +58,22 @@ const FilterModal: FunctionComponent<FilterModalProps> = ({
...
-    const [filterRules, setFilterRules, saveFilterRules, revertFilterRules, resetFilterRules] = useRevertableState<
-        FilterRule[]
-    >([]);
-    const [filterDirty, setFilterDirty, saveFilterDirty, revertFilterDirty, resetFilterDirty] = useRevertableState<
-        boolean
-    >(false);
+    const filterId = useRevertableState<string | undefined>(undefined);
+    const filterRules = useRevertableState<FilterRule[]>([]);
+    const filterDirty = useRevertableState<boolean>(false);
+    const modalState = useRevertableStates(filterId, filterRules, filterDirty);
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/header/FilterModal.tsx:64]: Nice :+1:

**Review by Gelio — APPROVED** [review]
LGTM :+1: A considerable improvement in terms of readability :smile:

File: widgets/common/src/deploymentsView/header/FilterModal.tsx (hunk 17 above)
Comment (Gelio) [reply widgets/common/src/deploymentsView/header/FilterModal.tsx]: Looks good, I did not think of `useRevertableStates`, but IMO that's a nice pattern :+1: I meant more of a hardcoded solution, but yours looks to scale better and is alright in terms of readability

**Comment by kubama** [comment]
Since I'm not receiving any comments from @stepek and this PR is open for a while I'm merging this PR with one approval.

@Gelio Thanks for spending your time on the review. Even though it was lengthy I think we managed to improve the quality drastically, so I think it was worth it.

**Review by stepek — COMMENTED** [review]
LGTM
